### PR TITLE
Parquet: Use new logical type annotations

### PR DESCRIPTION
**Pull request by rdblue**
Parquet 1.11.0 added logical type annotations that can be used to store Iceberg's `adjustToUTC` value like Avro, so that Parquet files correctly encode whether a value is a `timestamp` or a `timestamptz`.

### CODE REVIEW

**Comment by rdblue** [comment]
@Fokko, you may be interested in this one.

File: data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java [hunk 1]
```diff
@@ -200,6 +210,70 @@ private GenericParquetWriter() {
...
+    public Optional<PrimitiveWriter<?>> visit(LogicalTypeAnnotation.TimestampLogicalTypeAnnotation timestampType) {
+      if (timestampType.isAdjustedToUTC()) {
+        return Optional.of(new TimestamptzWriter(desc));
+      } else {
+        return Optional.of(new TimestampWriter(desc));
+      }
+    }
+
+    @Override
+    public Optional<PrimitiveWriter<?>> visit(LogicalTypeAnnotation.JsonLogicalTypeAnnotation jsonLogicalType) {
```
Comment (wangmiao1981) [inline data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java:267]: In Parquet, LogicalTypeAnnotation.java, there is `LogicalTypeAnnotation.IntLogicalTypeAnnotation`. Do we need it here too?

Comment (rdblue) [reply data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java:267]: Methods that aren't implemented here use the superclass implementation, which returns `Optional.empty`. When that happens, we fall back to the old logic that uses just the primitive type.

File: data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java [hunk 2]
```diff
@@ -117,6 +120,13 @@ private GenericParquetWriter() {
     @Override
     public ParquetValueWriter<?> primitive(PrimitiveType primitive) {
       ColumnDescriptor desc = type.getColumnDescription(currentPath());
+      LogicalTypeAnnotation logicalType = primitive.getLogicalTypeAnnotation();
+      if (logicalType != null) {
+        Optional<PrimitiveWriter<?>> writer = logicalType.accept(new LogicalTypeWriterVisitor(desc));
+        if (writer.isPresent()) {
+          return writer.get();
+        }
```
Comment (waterlx) [inline data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java:128]: When writer.isPresent() == false, we falls back to the previous logic, right?

Comment (rdblue) [reply data/src/main/java/org/apache/iceberg/data/parquet/GenericParquetWriter.java:128]: Correct.

**Comment by danielcweeks** [comment]
This all looks good to me. +1

**Comment by rdblue** [comment]
Thanks for reviewing, everyone!

**Pull request merged**